Sources/SwiftSyntax/SourceLocation.swift

```diff
@@ -478,57 +423,283 @@ extension SyntaxProtocol {
   }
 }
 
-/// Returns array of lines with the position at the start of the line and
-/// the end-of-file position.
-fileprivate func computeLines(
-  tree: Syntax
-) -> (
-  lines: [AbsolutePosition],
-  endOfFile: AbsolutePosition,
-  sourceLocationDirectives: [(sourceLine: Int, arguments: SourceLocationDirectiveArguments?)]
-) {
-  var lines: [AbsolutePosition] = [.startOfFile]
-  var position: AbsolutePosition = .startOfFile
-  var sourceLocationDirectives: [(sourceLine: Int, arguments: SourceLocationDirectiveArguments?)] = []
-  let lastLineLength = tree.raw.forEachLineLength { lineLength in
-    position += lineLength
-    lines.append(position)
-  } handleSourceLocationDirective: { lineOffset, args in
-    sourceLocationDirectives.append((sourceLine: lines.count + lineOffset, arguments: args))
-  }
-  return (lines, position + lastLineLength, sourceLocationDirectives)
+/// `Array` wrapper to perform binary searches.
+private struct SortedArray<Element>: RandomAccessCollection {
+  typealias Storage = [Element]
+  let storage: Storage
+
+  /// Initialize with the pre-sorted array.
+  ///
+  /// It is the caller's responsibility to pre-sort the array. Otherwise, the
+  /// behavior is undefined.
+  init(sortedArray: [Element]) {
+    self.storage = sortedArray
+  }
+
+  /// Find the first index where the element is ordered after `value`.
+  func upperBoundIndex<Value>(_ val: Value, by areIncreasingOrder: (Value, Element) -> Bool) -> Storage.Index {
+    var start = storage.startIndex
+    var end = storage.endIndex
+    while end > start {
+      let idx = (end - start) / 2 + start
+      if (!areIncreasingOrder(val, storage[idx])) {
+        start = idx + 1
+      } else {
+        end = idx
+      }
+    }
+    return start
+  }
+
+  /// Equivalent to `upperBoundIndex(val, by: <)`.
+  func upperBoundIndex(_ val: Element) -> Storage.Index where Element: Comparable {
+    self.upperBoundIndex(val, by: <)
+  }
+
+  var startIndex: Int { storage.startIndex }
+  var endIndex: Int { storage.endIndex }
+  subscript(index: Storage.Index) -> Element {
+    storage[index]
+  }
+  func makeIterator() -> Storage.Iterator {
+    storage.makeIterator()
+  }
+  func withContiguousStorageIfAvailable<R>(_ body: (UnsafeBufferPointer<Element>) throws -> R) rethrows -> R? {
+    try storage.withUnsafeBufferPointer(body)
+  }
+}
+
+extension SortedArray: Sendable where Element: Sendable {}
+
+/// Represent a virtual file region in the source file.
```

Comment:
`VirtualFile` corresponds to [`swift::SourceManager::VirtualFile`](https://github.com/swiftlang/swift/blob/25252cd6836f512ff820ff995ae8d92ab47076d1/include/swift/Basic/SourceManager.h#L130-L138)
This will be used for populating `swift::SourceManager`'s table see https://github.com/swiftlang/swift/pull/77721

Reply:
I think a doc comment would be helpful for this.